dev/PowerNotifications/PowerNotifications.cpp

```diff
@@ -0,0 +1,316 @@
+
+    // DisplayStatus Functions
+    EventType& DisplayStatus_Event()
+    {
+        return make_self<factory_implementation::PowerManager>()->m_displayStatusChangedEvent;
+    }
+
+    void DisplayStatus_Register()
+    {
+        check_hresult(PowerNotifications_RegisterDisplayStatusChangedListener(
+            &PowerManager::DisplayStatusChanged_Callback,
+            &make_self<factory_implementation::PowerManager>()->m_displayStatusHandle));
+    }
+
+    void DisplayStatus_Unregister()
+    {
+        check_hresult(PowerNotifications_UnregisterDisplayStatusChangedListener
+        (make_self<factory_implementation::PowerManager>()->m_displayStatusHandle));
+    }
+
+    void DisplayStatus_Update()
+    {
+        check_hresult(PowerNotifications_GetDisplayStatus(
+            &make_self<factory_implementation::PowerManager>()->m_cachedDisplayStatus));
+    }
+
+    // SystemIdleStatus Functions
+    EventType& SystemIdleStatus_Event()
+    {
+        return make_self<factory_implementation::PowerManager>()->m_systemIdleStatusChangedEvent;
+    }
+
+    void SystemIdleStatus_Register()
+    {
+        check_hresult(PowerNotifications_RegisterSystemIdleStatusChangedListener(
+            &PowerManager::SystemIdleStatusChanged_Callback,
+            &make_self<factory_implementation::PowerManager>()->m_systemIdleStatusHandle));
+    }
+
+    void SystemIdleStatus_Unregister()
+    {
+        check_hresult(PowerNotifications_UnregisterSystemIdleStatusChangedListener(
+            make_self<factory_implementation::PowerManager>()->m_systemIdleStatusHandle));
+    }
+
+    ULONG GetEffectivePowerModeVersion()
+    {
+        // We check for the version supported by checking if the feature is supported
+        // Using NULL as an indicator for uninitialized cache value
+        if (g_powerModeVersion == NULL)
+        {
+            PVOID handle;
+            auto hr = PowerRegisterForEffectivePowerModeNotifications(
+                EFFECTIVE_POWER_MODE_V2, [](EFFECTIVE_POWER_MODE, PVOID) {}, NULL, &handle);
+            g_powerModeVersion = hr == S_OK ? EFFECTIVE_POWER_MODE_V2 : EFFECTIVE_POWER_MODE_V1;
+            check_hresult(PowerUnregisterFromEffectivePowerModeNotifications(handle));
+        }
+        return g_powerModeVersion;
+    }
+
+    EventType& EffectivePowerMode_Event()
+    {
+        return make_self<factory_implementation::PowerManager>()->m_powerModeChangedEvent;
+    }
+
+    void EffectivePowerMode_Register()
+    {
+        ULONG version = GetEffectivePowerModeVersion();
+        check_hresult(PowerRegisterForEffectivePowerModeNotifications(
+            version, [](EFFECTIVE_POWER_MODE mode, PVOID)
+            {
+                PowerManager::EffectivePowerModeChanged_Callback(mode);
+            }, nullptr, &g_powerModeHandle));
+    }
+
+    void EffectivePowerMode_Unregister()
+    {
+        check_hresult(PowerUnregisterFromEffectivePowerModeNotifications(g_powerModeHandle));
+        g_powerModeHandle = NULL;
```

Comment:
I'm getting code smell from the global.  Is a global truly needed?

Reply:
Changed to member.